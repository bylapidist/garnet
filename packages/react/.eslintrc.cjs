module.exports = {
  root: true,
  extends: ['@lapidist/garnet'],
};
