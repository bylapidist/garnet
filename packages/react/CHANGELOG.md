# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.5.48](https://github.com/bylapidist/garnet/compare/v0.5.47...v0.5.48) (2024-03-17)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.5.45](https://github.com/bylapidist/garnet/compare/v0.5.44...v0.5.45) (2024-03-13)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.5.41](https://github.com/bylapidist/garnet/compare/v0.5.40...v0.5.41) (2024-03-07)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.5.36](https://github.com/bylapidist/garnet/compare/v0.5.35...v0.5.36) (2024-03-02)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.5.33](https://github.com/bylapidist/garnet/compare/v0.5.32...v0.5.33) (2024-02-20)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.5.29](https://github.com/bylapidist/garnet/compare/v0.5.28...v0.5.29) (2024-02-15)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.5.22](https://github.com/bylapidist/garnet/compare/v0.5.21...v0.5.22) (2024-02-07)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.5.17](https://github.com/bylapidist/garnet/compare/v0.5.16...v0.5.17) (2024-02-06)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.5.4](https://github.com/bylapidist/garnet/compare/v0.5.3...v0.5.4) (2024-01-18)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.5.2](https://github.com/bylapidist/garnet/compare/v0.5.1...v0.5.2) (2024-01-18)

**Note:** Version bump only for package @lapidist/garnet-react





# [0.5.0](https://github.com/bylapidist/garnet/compare/v0.4.23...v0.5.0) (2024-01-16)


### Features

* **apps/website:** add nextjs scaffolding to apps/website package ([ffca25b](https://github.com/bylapidist/garnet/commit/ffca25b89d56e470543b92cf3739fe3ce31c738e))





## [0.4.20](https://github.com/bylapidist/garnet/compare/v0.4.19...v0.4.20) (2024-01-12)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.4.15](https://github.com/bylapidist/garnet/compare/v0.4.14...v0.4.15) (2024-01-05)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.4.5](https://github.com/bylapidist/garnet/compare/v0.4.4...v0.4.5) (2023-12-17)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.4.4](https://github.com/bylapidist/garnet/compare/v0.4.3...v0.4.4) (2023-12-16)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.4.3](https://github.com/bylapidist/garnet/compare/v0.4.2...v0.4.3) (2023-12-16)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.4.2](https://github.com/bylapidist/garnet/compare/v0.4.1...v0.4.2) (2023-12-16)


### Bug Fixes

* missing garnet-styles dependency ([3c2d6de](https://github.com/bylapidist/garnet/commit/3c2d6de947e6250aac3eb8b7b1382ec126196bbd))





# [0.4.0](https://github.com/bylapidist/garnet/compare/v0.3.4...v0.4.0) (2023-12-16)


### Features

* **styles:** implement styles and hot reloading ([3cc4cac](https://github.com/bylapidist/garnet/commit/3cc4cacb42ab61bbc43ee9b614c00705b113c46a))





## [0.3.4](https://github.com/bylapidist/garnet/compare/v0.3.3...v0.3.4) (2023-12-16)

**Note:** Version bump only for package @lapidist/garnet-react





# [0.3.0](https://github.com/bylapidist/garnet/compare/v0.2.23...v0.3.0) (2023-12-13)


### Features

* **tokens:** add tokens package ([53f13b3](https://github.com/bylapidist/garnet/commit/53f13b310df90fd37cd811a0a142efb3500bcb55))





## [0.2.3](https://github.com/bylapidist/garnet/compare/v0.2.2...v0.2.3) (2023-08-27)

**Note:** Version bump only for package @lapidist/garnet-react





# [0.2.0](https://github.com/bylapidist/garnet/compare/v0.1.2...v0.2.0) (2023-08-26)


### Features

* **storybook:** add storybook app ([b88dc90](https://github.com/bylapidist/garnet/commit/b88dc9008cd8ed325c96ee911e5209411b27994f))





## [0.1.2](https://github.com/bylapidist/garnet/compare/v0.1.1...v0.1.2) (2023-08-25)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.1.1](https://github.com/bylapidist/garnet/compare/v0.1.1-beta.1...v0.1.1) (2023-08-24)

**Note:** Version bump only for package @lapidist/garnet-react





## [0.1.1-beta.1](https://github.com/bylapidist/garnet/compare/v0.1.1-beta.0...v0.1.1-beta.1) (2023-08-24)


### Bug Fixes

* **react:** add button type ([1c10ed5](https://github.com/bylapidist/garnet/commit/1c10ed5565034e6a4e1fb7bc3753707012eaafed))





## [0.1.1-beta.0](https://github.com/bylapidist/garnet/compare/v0.1.0...v0.1.1-beta.0) (2023-08-24)

**Note:** Version bump only for package @lapidist/garnet-react





# [0.1.0](https://github.com/bylapidist/garnet/compare/v0.1.0-beta.0...v0.1.0) (2023-08-24)

**Note:** Version bump only for package @lapidist/garnet-react





# 0.1.0-beta.0 (2023-08-24)

**Note:** Version bump only for package @lapidist/garnet-react
