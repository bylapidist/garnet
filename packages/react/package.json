{
  "name": "@lapidist/garnet-react",
  "version": "0.5.48",
  "description": "Lapidist components",
  "author": {
    "name": "Brett Dorrans",
    "email": "hello@lapidist.net"
  },
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/bylapidist/garnet.git"
  },
  "bugs": {
    "url": "https://github.com/bylapidist/garnet/issues"
  },
  "homepage": "https://github.com/bylapidist/garnet#readme",
  "publishConfig": {
    "access": "public"
  },
  "keywords": [
    "react",
    "garnet",
    "lapidist",
    "components",
    "component library",
    "design system"
  ],
  "type": "module",
  "files": [
    "dist",
    "CHANGELOG.md"
  ],
  "main": "./dist/index.umd.js",
  "module": "./dist/index.es.js",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "require": "./dist/index.umd.js",
      "import": "./dist/index.es.js"
    },
    "./style.css": {
      "import": "./dist/style.css",
      "require": "./dist/style.css"
    }
  },
  "scripts": {
    "build": "tsc && vite build",
    "watch": "concurrently npm:watch:ts npm:watch:vite",
    "watch:ts": "tsc -w",
    "watch:vite": "vite build --watch",
    "watch:tests": "vitest",
    "lint": "eslint .",
    "fix": "eslint . --fix",
    "test": "vitest run --coverage",
    "typecheck": "tsc --noEmit -p tsconfig.json",
    "snapshots": "vitest run --update"
  },
  "dependencies": {
    "@lapidist/garnet-styles": "^0.5.48",
    "@lapidist/garnet-tokens": "^0.5.4",
    "react": "18.x",
    "react-dom": "18.x"
  },
  "devDependencies": {
    "@lapidist/garnet-types": "^0.5.0"
  },
  "peerDependencies": {
    "react": "18.x",
    "react-dom": "18.x"
  }
}
