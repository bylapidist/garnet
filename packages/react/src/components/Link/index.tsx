import React from 'react';
import cx from 'classnames';
import { CommonProps, LinkProps } from '@lapidist/garnet-types';

import styles from '@lapidist/garnet-styles/components/link.module.css';
import Text from '../Text';
import Box from '../Box';

export default function Link({
    id,
    className,
    as = 'a',
    testId = 'Link',
    size = 'medium',
    family = 'body',
    weight = 'regular',
    children,
    ...restProps
}: React.PropsWithChildren<CommonProps<LinkProps>>) {
    return (
        <Text
            as={as}
            testId={testId}
            id={id}
            className={cx(styles.this, className)}
            size={size}
            family={family}
            weight={weight}
            {...restProps}
        >
            <Box as="span" testId={`${testId}-inner`} className={styles.inner}>
                {children}
            </Box>
        </Text>
    );
}
