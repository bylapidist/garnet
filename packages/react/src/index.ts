export { default as App } from './components/App';
export { default as Button } from './components/Button';
export { default as Text } from './components/Text';
