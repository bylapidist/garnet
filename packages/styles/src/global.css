html, body, div, span, applet, object, iframe,
h1, h2, h3, h4, h5, h6, p, blockquote, pre,
a, abbr, acronym, address, big, cite, code,
del, dfn, em, img, ins, kbd, q, s, samp,
small, strike, strong, sub, sup, tt, var,
b, u, i, center, button,
dl, dt, dd, ol, ul, li,
fieldset, form, label, legend,
table, caption, tbody, tfoot, thead, tr, th, td,
article, aside, canvas, details, embed,
figure, figcaption, footer, header, hgroup,
main, menu, nav, output, ruby, section, summary,
time, mark, audio, video {
    margin: 0;
    padding: 0;
    border: 0;
    font: inherit;
    font-size: 100%;
    vertical-align: baseline;
    transition: border var(--transition-ease-in-fast),
    background-color var(--transition-ease-in-fast),
    text-decoration-color var(--transition-ease-in-fast);
}

html {
    box-sizing: border-box;
}

*,
*::before,
*::after {
    box-sizing: inherit;
}

article,
aside,
details,
figcaption,
figure,
footer,
header,
hgroup,
main,
menu,
nav,
section {
    display: block;
}

*[hidden] {
    display: none;
}

body {
    line-height: 1;
    font-size: 62.5%;
}

menu,
ol,
ul {
    list-style: none;
}

blockquote,
q {
    quotes: none;
}

blockquote::before,
blockquote::after,
q::before,
q::after {
    content: '';
}

table {
    border-collapse: collapse;
    border-spacing: 0;
}

img {
    display: block;
    max-width: 100%;
}

a,
button,
input {
    text-decoration: none;
    outline-offset: var(--border-width-thinner);
    outline: transparent solid var(--border-width-thinner);
}

a:focus,
button:focus,
input:focus {
    outline: var(--colour-secondary-dark) solid var(--border-width-thinner);
}
