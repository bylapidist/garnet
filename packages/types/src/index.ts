export type { ButtonProps } from './components/Button';
