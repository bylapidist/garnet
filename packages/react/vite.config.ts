/// <reference types="vitest" />
import react from '@vitejs/plugin-react';
import path from 'path';
import { defineConfig } from 'vite';
import dts from 'vite-plugin-dts';

// https://vitejs.dev/config/
export default defineConfig({
    plugins: [
        react(),
        dts({
            insertTypesEntry: true,
        }),
    ],
    test: {
        globals: true,
        environment: 'jsdom',
        coverage: {
            provider: 'v8',
        },
    },
    build: {
        lib: {
            entry: path.resolve(__dirname, 'src/index.ts'),
            name: 'garnet-react',
            formats: ['es', 'umd'],
            fileName: (format) => `index.${format}.js`,
        },
        rollupOptions: {
            external: ['react', 'react-dom'],
            output: {
                globals: {
                    react: 'React',
                    'react-dom': 'ReactDOM',
                },
            },
        },
    },
});
