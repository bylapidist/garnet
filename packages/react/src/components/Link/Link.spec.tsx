import React from 'react';
import { describe, expect, test, afterEach } from 'vitest';
import { cleanup, render, screen } from '@testing-library/react';
import Link from './index';
import {
    TextFamily,
    TextWeight,
} from '@lapidist/garnet-types/src/components/Text';
import { Size } from '@lapidist/garnet-types/src/common';

afterEach(cleanup);

const setup = (Component: React.ReactElement) => render(Component);

const sizes: Size[] = ['small', 'medium', 'large'];
const families: TextFamily[] = ['body', 'heading', 'serif', 'mono'];
const weights: TextWeight[] = ['regular', 'medium', 'bold'];

describe('Link', () => {
    test('it works with defaults', () => {
        const { container } = setup(<Link href="#">Hello world</Link>);
        expect(container.firstChild).toMatchSnapshot();
    });

    test.each(sizes)('it works with sizes', (size) => {
        const { container } = setup(
            <Link href="#" size={size}>
                Hello world
            </Link>,
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    test.each(families)('it works with families', (family) => {
        const { container } = setup(<Link family={family}>Hello world</Link>);
        expect(container.firstChild).toMatchSnapshot();
    });

    test.each(weights)('it works with weights', (weight) => {
        const { container } = setup(<Link weight={weight}>Hello world</Link>);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('it works with default testId', () => {
        setup(<Link>Hello world</Link>);
        expect(screen.getByTestId('Link')).toBeTruthy();
    });

    test('it works with specified testId', () => {
        setup(<Link testId={'TestId'}>Hello world</Link>);
        expect(screen.getByTestId('TestId')).toBeTruthy();
    });

    test('it works with as', () => {
        setup(<Link as="button">Hello world</Link>);
        expect(screen.getByTestId('Link').nodeName).toBe('BUTTON');
    });

    test('it works with className', () => {
        setup(<Link className="test">Hello world</Link>);
        expect(
            screen.getByTestId('Link').classList.contains('test'),
        ).toBeTruthy();
    });
});
