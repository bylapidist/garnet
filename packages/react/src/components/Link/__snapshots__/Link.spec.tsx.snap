// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link > it works with defaults 1`] = `
<a
  class="_this_36ca96 _size-medium_36ca96 _family-body_36ca96 _weight-regular_36ca96 _this_330137"
  data-testid="Link"
  href="#"
>
  <span
    class="_elevation-none_02cc31 _roundness-none_02cc31 _gutter-none_02cc31 _inner_330137"
    data-testid="Link-inner"
  >
    Hello world
  </span>
</a>
`;

exports[`Link > it works with families 1`] = `
<a
  class="_this_36ca96 _size-medium_36ca96 _family-body_36ca96 _weight-regular_36ca96 _this_330137"
  data-testid="Link"
>
  <span
    class="_elevation-none_02cc31 _roundness-none_02cc31 _gutter-none_02cc31 _inner_330137"
    data-testid="Link-inner"
  >
    Hello world
  </span>
</a>
`;

exports[`Link > it works with families 2`] = `
<a
  class="_this_36ca96 _size-medium_36ca96 _family-heading_36ca96 _weight-regular_36ca96 _this_330137"
  data-testid="Link"
>
  <span
    class="_elevation-none_02cc31 _roundness-none_02cc31 _gutter-none_02cc31 _inner_330137"
    data-testid="Link-inner"
  >
    Hello world
  </span>
</a>
`;

exports[`Link > it works with families 3`] = `
<a
  class="_this_36ca96 _size-medium_36ca96 _family-serif_36ca96 _weight-regular_36ca96 _this_330137"
  data-testid="Link"
>
  <span
    class="_elevation-none_02cc31 _roundness-none_02cc31 _gutter-none_02cc31 _inner_330137"
    data-testid="Link-inner"
  >
    Hello world
  </span>
</a>
`;

exports[`Link > it works with families 4`] = `
<a
  class="_this_36ca96 _size-medium_36ca96 _family-mono_36ca96 _weight-regular_36ca96 _this_330137"
  data-testid="Link"
>
  <span
    class="_elevation-none_02cc31 _roundness-none_02cc31 _gutter-none_02cc31 _inner_330137"
    data-testid="Link-inner"
  >
    Hello world
  </span>
</a>
`;

exports[`Link > it works with sizes 1`] = `
<a
  class="_this_36ca96 _size-small_36ca96 _family-body_36ca96 _weight-regular_36ca96 _this_330137"
  data-testid="Link"
  href="#"
>
  <span
    class="_elevation-none_02cc31 _roundness-none_02cc31 _gutter-none_02cc31 _inner_330137"
    data-testid="Link-inner"
  >
    Hello world
  </span>
</a>
`;

exports[`Link > it works with sizes 2`] = `
<a
  class="_this_36ca96 _size-medium_36ca96 _family-body_36ca96 _weight-regular_36ca96 _this_330137"
  data-testid="Link"
  href="#"
>
  <span
    class="_elevation-none_02cc31 _roundness-none_02cc31 _gutter-none_02cc31 _inner_330137"
    data-testid="Link-inner"
  >
    Hello world
  </span>
</a>
`;

exports[`Link > it works with sizes 3`] = `
<a
  class="_this_36ca96 _size-large_36ca96 _family-body_36ca96 _weight-regular_36ca96 _this_330137"
  data-testid="Link"
  href="#"
>
  <span
    class="_elevation-none_02cc31 _roundness-none_02cc31 _gutter-none_02cc31 _inner_330137"
    data-testid="Link-inner"
  >
    Hello world
  </span>
</a>
`;

exports[`Link > it works with weights 1`] = `
<a
  class="_this_36ca96 _size-medium_36ca96 _family-body_36ca96 _weight-regular_36ca96 _this_330137"
  data-testid="Link"
>
  <span
    class="_elevation-none_02cc31 _roundness-none_02cc31 _gutter-none_02cc31 _inner_330137"
    data-testid="Link-inner"
  >
    Hello world
  </span>
</a>
`;

exports[`Link > it works with weights 2`] = `
<a
  class="_this_36ca96 _size-medium_36ca96 _family-body_36ca96 _weight-medium_36ca96 _this_330137"
  data-testid="Link"
>
  <span
    class="_elevation-none_02cc31 _roundness-none_02cc31 _gutter-none_02cc31 _inner_330137"
    data-testid="Link-inner"
  >
    Hello world
  </span>
</a>
`;

exports[`Link > it works with weights 3`] = `
<a
  class="_this_36ca96 _size-medium_36ca96 _family-body_36ca96 _weight-bold_36ca96 _this_330137"
  data-testid="Link"
>
  <span
    class="_elevation-none_02cc31 _roundness-none_02cc31 _gutter-none_02cc31 _inner_330137"
    data-testid="Link-inner"
  >
    Hello world
  </span>
</a>
`;
